### PR TITLE
mon: fix/improve mon sync over small keys

### PR DESCRIPTION
**Pull request by liewegas**
A user experienced mon elections/timeouts trying to sync over small keys.

1. Make sync more CPU efficient
2. Cap messages to 2000 keys

### CODE REVIEW

File: src/common/legacy_config_opts.h [hunk 1]
```diff
@@ -274,6 +274,7 @@ OPTION(mon_scrub_inject_missing_keys, OPT_DOUBLE) // probability of injected mis
 OPTION(mon_config_key_max_entry_size, OPT_INT) // max num bytes per config-key entry
 OPTION(mon_sync_timeout, OPT_DOUBLE)
 OPTION(mon_sync_max_payload_size, OPT_U32) // max size for a sync chunk payload (say)
+OPTION(mon_sync_max_payload_keys, OPT_U32) // max size for a sync chunk payload (say)
```
Comment (tchaikov) [inline src/common/legacy_config_opts.h]: s/max size/max number of keys/

Comment (tchaikov) [inline src/common/legacy_config_opts.h]: s/max size/max number of keys/

File: src/mon/MonitorDBStore.h [hunk 2]
```diff
@@ -481,17 +458,35 @@ class MonitorDBStore
...
       ceph_assert(iter->valid() == true);
 
-      while (iter->valid()) {
+      while (iter->valid() && max_bytes > 0) {
 	string prefix(iter->raw_key().first);
 	string key(iter->raw_key().second);
 	if (sync_prefixes.count(prefix)) {
 	  bufferlist value = iter->value();
-	  if (!add_chunk_entry(tx, prefix, key, value, max))
+	  if (!tx->empty() &&
```
Comment (tchaikov) [inline src/mon/MonitorDBStore.h]: i don't think this is correct. what if tx is empty at this moment?
shall we instead do something like
```c++
if (tx->get_bytes() + value.length() + key.size() +
    prefix.size() < max_bytes &&
    tx->get_keys() < max_keys) {
  // ...
}
```

as in http://pulpito.ceph.com/kchai-2019-11-15_17:09:33-rados-wip-kefu-testing-2019-11-13-1728-distro-basic-mira/, i found that lots of occurrences of "failed to reach quorum size 3 before timeout expired".

after checking the log, it turns out the mon which failed to join the quorum was keeping asking for its sync provider for the same paxos chunk.

see 
```
zgrep 'through version' /a/kchai-2019-11-15_17:09:33-rados-wip-kefu-testing-2019-11-13-1728-distro-basic-mira/4510653/remote/mira069/log/ceph-mon.a.log.gz | head -n100
```

and the provide kept sending empty mon sync replies
```
2019-11-15T17:38:31.670+0000 7fb2ed5cb700  1 -- [v2:172.21.6.110:3300/0,v1:172.21.6.110:6789/0] --> [v2:172.21.6.110:3301/0,v1:172.21.6.110:6790/0] -- mon_sync(chunk cookie 1006632961 lc 124 bl 26 bytes) v2 -- 0x557fc4f6b8c0 con 0x557fc37d3a80
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by liewegas** [comment]
passing tests now. i got the legacy_config_opts.h type wrong with was screwing up the per-key limit.

**Pull request merged**